File: pr1-java/src/main/java/repository/database/UserDbRepository.java
package repository.database;

import domain.User;
import domain.exceptions.DuplicateException;
import domain.exceptions.NotFoundException;
import repository.UserRepository;
import utils.Configuration;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class UserDbRepository implements UserRepository {
    public UserDbRepository() {
        Configuration.logger.info("Initializing UserDbRepository with {} ", Configuration.properties);
    }

    @Override
    public Iterable<User> getAll() {
        Configuration.logger.traceEntry("Getting all users");

        Connection connection = DbUtils.getConnection();
        List<User> users = new ArrayList<>();
        try (ResultSet result = connection.prepareStatement("select * from users;").executeQuery()) {
            while (result.next()) {
                Integer userId = result.getInt("userId");
                String username = result.getString("username");
                String status = result.getString("status");
                User user = new User(userId, username, status);
                users.add(user);
            }
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(users);
        return users;
    }

    @Override
    public User getOne(Integer id) throws NotFoundException {
        Configuration.logger.traceEntry("Getting user with {}", id);

        Connection connection = DbUtils.getConnection();
        try (ResultSet result = connection.prepareStatement("select * from users where userId = " + id + ";").executeQuery()) {
            String username = result.getString("username");
            String status = result.getString("status");
            User user = new User(id, username, status);

            Configuration.logger.traceExit(user);
            return user;
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }
    }

    @Override
    public User getOne(String username) throws NotFoundException {
        Configuration.logger.traceEntry("Getting user with {}", username);

        Connection connection = DbUtils.getConnection();
        try (ResultSet result = connection.prepareStatement("select * from users where username = '" + username + "';").executeQuery()) {
            Integer userId = result.getInt("userId");
            String status = result.getString("status");
            User user = new User(userId, username, status);

            Configuration.logger.traceExit(user);
            return user;
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }
    }

    @Override
    public User add(User user) throws DuplicateException {
        Configuration.logger.traceEntry("Saving {} ", user);

        try {
            getOne(user.getUsername());
            throw new DuplicateException();
        } catch (NotFoundException ignored) {
        }

        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into users(username, status) values(?,?);")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getStatus());

            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Saved {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        try (ResultSet result = connection.prepareStatement("select userId from users where username = '" + user.getUsername() + "';").executeQuery()) {
            user.setId(result.getInt("userId"));
            Configuration.logger.trace("Got id for {}", user);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(user);
        return user;
    }

    @Override
    public User remove(Integer id) throws NotFoundException {
        Configuration.logger.traceEntry("Removing user with {} ", id);

        User user = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from users where userId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Deleted {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }

        Configuration.logger.traceExit(user);
        return user;
    }

    @Override
    public User modify(Integer id, User newUser) throws NotFoundException {
        Configuration.logger.traceEntry("Modifying user with {} ", id);

        User oldUser = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update users set " +
                "username = '" + newUser.getUsername() +
                "', status = '" + newUser.getStatus() +
                "' where userId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Modified {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(oldUser);
        return oldUser;
    }

    @Override
    public User setUserStatus(Integer id, String status) throws NotFoundException {
        Configuration.logger.traceEntry("Setting status for user with {} ", id);

        User user = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update users set status = '" + status + "' where userId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Status set for {}", user);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(user);
        return user;
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        for (User user : getAll())
            string.append(user).append("\n");
        return string.toString();
    }
}

File: pr1-cs/Repository/Database/UserDbRepository.cs
using System.Collections.Generic;
using log4net;
using pr1_cs.Domain;
using pr1_cs.Domain.Exceptions;

namespace pr1_cs.Repository.Database
{
    public class UserDbRepository : IUserRepository
    {
        private static readonly ILog Log = LogManager.GetLogger("UserDbRepository");

        public UserDbRepository()
        {
            Log.Info("Creating UserDbRepository");
        }

        public IEnumerable<User> GetAll()
        {
            Log.Info("Entering GetAll");

            var users = new List<User>();
            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from users;";
                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        var userId = dataReader.GetInt32(dataReader.GetOrdinal("userId"));
                        var username = dataReader.GetString(dataReader.GetOrdinal("username"));
                        var status = dataReader.GetString(dataReader.GetOrdinal("status"));

                        var user = new User(userId, username, status);
                        users.Add(user);
                    }
                }
            }

            Log.InfoFormat("Exiting GetAll with values {0}", users);
            return users;
        }

        public User GetOne(int id)
        {
            Log.InfoFormat("Entering GetOne with value {0}", id);

            User user;
            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from users where userId = " + id + ";";
                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        var username = dataReader.GetString(dataReader.GetOrdinal("username"));
                        var status = dataReader.GetString(dataReader.GetOrdinal("status"));

                        user = new User(id, username, status);
                    }
                    else
                    {
                        throw new NotFoundException();
                    }
                }
            }

            Log.InfoFormat("Exiting GetOne with value {0}", user);
            return user;
        }

        public User GetOne(string username)
        {
            Log.InfoFormat("Entering GetOne with value {0}", username);

            User user;
            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from users where username = '" + username + "';";
                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        var userId = dataReader.GetInt32(dataReader.GetOrdinal("userId"));
                        var status = dataReader.GetString(dataReader.GetOrdinal("status"));

                        user = new User(userId, username, status);
                    }
                    else
                    {
                        throw new NotFoundException();
                    }
                }
            }

            Log.InfoFormat("Exiting GetOne with value {0}", user);
            return user;
        }

        public User Add(User user)
        {
            Log.InfoFormat("Entering Add with value {0}", user);

            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "insert into users(username, status) values(@username, @status);";

                var dataParameter = command.CreateParameter();
                dataParameter.ParameterName = "@username";
                dataParameter.Value = user.Username;
                command.Parameters.Add(dataParameter);

                dataParameter = command.CreateParameter();
                dataParameter.ParameterName = "@status";
                dataParameter.Value = user.Status;
                command.Parameters.Add(dataParameter);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new DuplicateException();

                command.CommandText = "select userId from users where username = '" + user.Username + "';";
                using (var dataReader = command.ExecuteReader())
                {
                    dataReader.Read();
                    var userId = dataReader.GetInt32(dataReader.GetOrdinal("userId"));
                    user.Id = userId;
                }
            }

            Log.InfoFormat("Exiting Add with value {0}", user);
            return user;
        }

        public User Remove(int id)
        {
            Log.InfoFormat("Entering Remove with value {0}", id);

            var user = GetOne(id);
            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "delete from users where userId = " + id + ";";
                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new NotFoundException();
            }

            Log.InfoFormat("Exiting Remove with value {0}", user);
            return user;
        }

        public User Modify(int id, User newUser)
        {
            Log.InfoFormat("Entering Modify with value {0}", id);

            var oldUser = GetOne(id);
            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "update users set " +
                                      "username = '" + newUser.Username +
                                      "', status = '" + newUser.Status +
                                      "' where userId = " + id + ";";
                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new NotFoundException();
            }

            Log.InfoFormat("Exiting Remove with value {0}", oldUser);
            return oldUser;
        }

        public User SetUserStatus(int id, string status)
        {
            Log.InfoFormat("Entering SetUserStatus with value {0}", id);

            var user = GetOne(id);
            var connection = DbUtils.Connection;
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "update users set status = '" + status + "' where userId = " + id + ";";
                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new NotFoundException();
            }

            Log.InfoFormat("Exiting SetUserStatus with value {0}", user);
            return user;
        }

        public override string ToString()
        {
            var stringRepo = "";
            foreach (var user in GetAll()) stringRepo += user + "\n";
            return stringRepo;
        }
    }
}